spark3-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMergeIntoTable.java

```diff
@@ -0,0 +1,275 @@
+    append(sourceName, new Employee(2, "emp-id-2"), new Employee(1, "emp-id-1"), new Employee(6, "emp-id-6"));
+    String sqlText = "MERGE INTO %s AS target " +
+            "USING %s AS source " +
+            "ON target.id = source.id " +
+            "WHEN MATCHED AND target.id = 6 THEN DELETE";
+
+    sql(sqlText, targetName, sourceName);
+    assertEquals("Should have expected rows",
+            ImmutableList.of(row(1, "emp-id-one")),
+            sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
+  }
+
+  @Test
+  public void testAllCauses() throws NoSuchTableException {
+    append(targetName, new Employee(1, "emp-id-one"), new Employee(6, "emp-id-6"));
+    append(sourceName, new Employee(2, "emp-id-2"), new Employee(1, "emp-id-1"), new Employee(6, "emp-id-6"));
+    String sqlText = "MERGE INTO %s AS target " +
+                     "USING %s AS source " +
+                     "ON target.id = source.id " +
+                     "WHEN MATCHED AND target.id = 1 THEN UPDATE SET * " +
+                     "WHEN MATCHED AND target.id = 6 THEN DELETE " +
+                     "WHEN NOT MATCHED AND source.id = 2 THEN INSERT * ";
+
+    sql(sqlText, targetName, sourceName);
+    assertEquals("Should have expected rows",
+            ImmutableList.of(row(1, "emp-id-1"), row(2, "emp-id-2")),
+            sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
+  }
+
+  @Test
+  public void testAllCausesWithExplicitColumnSpecification() throws NoSuchTableException {
+    append(targetName, new Employee(1, "emp-id-one"), new Employee(6, "emp-id-6"));
+    append(sourceName, new Employee(2, "emp-id-2"), new Employee(1, "emp-id-1"), new Employee(6, "emp-id-6"));
+    String sqlText = "MERGE INTO %s AS target " +
+            "USING %s AS source " +
+            "ON target.id = source.id " +
+            "WHEN MATCHED AND target.id = 1 THEN UPDATE SET target.id = source.id, target.dep = source.dep " +
+            "WHEN MATCHED AND target.id = 6 THEN DELETE " +
+            "WHEN NOT MATCHED AND source.id = 2 THEN INSERT (target.id, target.dep) VALUES (source.id, source.dep) ";
+
+    sql(sqlText, targetName, sourceName);
+    assertEquals("Should have expected rows",
+            ImmutableList.of(row(1, "emp-id-1"), row(2, "emp-id-2")),
+            sql("SELECT * FROM %s ORDER BY id ASC NULLS LAST", targetName));
+  }
+
+  @Test
+  public void testSourceCTE() throws NoSuchTableException {
+    Assume.assumeFalse(catalogName.equalsIgnoreCase("testhadoop"));
+    Assume.assumeFalse(catalogName.equalsIgnoreCase("testhive"));
+
+    append(targetName, new Employee(2, "emp-id-two"), new Employee(6, "emp-id-6"));
+    append(sourceName, new Employee(2, "emp-id-3"), new Employee(1, "emp-id-2"), new Employee(5, "emp-id-6"));
+    String sourceCTE = "WITH cte1 AS (SELECT id + 1 AS id, dep FROM source)";
+    String sqlText = sourceCTE + " " + "MERGE INTO %s AS target " +
```

Comment:
Nit: it looks like there are unnecessary string literals. `" " + "MERGE ..."` can be updated to `" MERGE ..."`.